File: code_snippets/ts/src/use_cases/async_tasks/sync_to_async/data_preparation_service.ts
import * as restate from "@restatedev/restate-sdk";
import {WorkflowContext, WorkflowSharedContext} from "@restatedev/restate-sdk";

// <start_here>
const dataPreparationService = restate.workflow({
    name: "dataPrep",
    handlers: {
        run: async (ctx: WorkflowContext, args: { userId: string }) => {
            // <mark_1>
            const url = await ctx.run("create S3 bucket", () => createS3Bucket());
            await ctx.run("upload data", () => uploadData(url));

            // <mark_2>
            await ctx.promise<URL>("url").resolve(url);
            // </mark_2>

            return url;
            // </mark_1>
        },

        resultAsEmail: async (ctx: WorkflowSharedContext, req: { email: string }) => {
            // <mark_2>
            const url = await ctx.promise<URL>("url");
            // </mark_2>
            await ctx.run("send email", () => sendEmail(url, req.email ));
        }
    }
});

export type DataPrepService = typeof dataPreparationService;
// <end_here>

async function createS3Bucket(): Promise<URL> {
    const bucket = Number(Math.random() * 1_000_000_000).toString(16);
    return new URL(`https://s3-eu-central-1.amazonaws.com/${bucket}/`);
}

async function uploadData(target: URL) {
    // simulate some work by delaying for a while. sometimes takes really long.
    return new Promise((resolve) => setTimeout(resolve, Math.random() < 0.0 ? 1_500 : 10_000));
}

async function sendEmail(url: URL, email: string) {
    // send email
    console.log(` >>> Sending email to '${email}' with URL ${url}`);
}

File: code_snippets/python/src/use_cases/async_tasks/sync_to_async/data_preparation_service.py
from restate import WorkflowContext, WorkflowSharedContext, Workflow
from typing import TypedDict


class DataPrepService(TypedDict):
    name: str
    handlers: dict


def create_s3_bucket() -> str:
    pass


def upload_data(target: str):
    pass


def send_email(url: str, email: str):
    print(f" >>> Sending email to '{email}' with URL {url}")


# <start_here>
data_preparation = Workflow("dataPrep")


@data_preparation.main()
async def run(ctx: WorkflowContext) -> str:
    # <mark_1>
    url = await ctx.run("bucket creation", lambda: create_s3_bucket())
    await ctx.run("upload", lambda: upload_data(url))

    # <mark_2>
    await ctx.promise("url").resolve(url)
    # </mark_2>

    return url
    # </mark_1>


@data_preparation.handler()
async def result_as_email(ctx: WorkflowSharedContext, email: str):
    # <mark_2>
    url = await ctx.promise("url").value()
    # </mark_2>
    await ctx.run("email", lambda: send_email(url, email))
# <end_here>
